deploy/infra/lib/helpers/base-lambda.ts

```diff
@@ -0,0 +1,28 @@
+import * as cdk from "aws-cdk-lib";
+import * as lambda from "aws-cdk-lib/aws-lambda";
+import { PolicyStatement } from "aws-cdk-lib/aws-iam";
+
+export interface BaseLambdaFunProps {
+  functionProps: lambda.FunctionProps;
+  vpcConfig: any;
+}
+
+export class BaseLambdaFunction extends lambda.Function {
+  constructor(scope: any, id: string, props: BaseLambdaFunProps) {
+    super(scope, id, props.functionProps);
+    this.addToRolePolicy(
```

Comment:
In cases where users aren't using VPCs with their deployment this will add unnecessary permissions - could you please attach [a `CfnCondition`](https://loige.co/create-resources-conditionally-with-cdk/#using-cfncondition-with-cdk) to this to only attach these permissions if the VPC is in use?

(unfortunately the way we render the CDK code into CloudFormation means that we can't do this at synthesis time which adds a bit of complexity. Usually you would just add some TypeScript code to conditionally add the PolicyStatement)

Reply:
Great point @chrnorm! Thank you for pointing it out!

Regarding the link you provided: As far as I understand it is for L1 resources ([link](https://docs.aws.amazon.com/cdk/v2/guide/constructs.html). However in our case it's a property that we add to a resource(namely IAM role which is created by default if you don't provide it to a constructor of lambda object). As you said we can't use runtime if statements because we do synthesis.

Some facts: A lambda function has only one execution role. Role is a high level L1 resources in terms of CF, so we can override different properties of it using structure like this:
```
    const cfnDist = resource.node.defaultChild as cdk.CfnResource;
    cfnDist.addPropertyOverride("PropertyName", "PropertyVal");
```
A lambda function is also a resource where one of the property is lambda function.

Possible solution: we can refactor code and create an execution lambda role ourselves rather than calling `addToRolePolicy` method. In this case we can do some "if/else" constructions because IAM Policy is construct. 
```
// Create a new IAM role
const myRole = new iam.Role(this, 'MyRole', {
  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com')
});

// Create policy
const policyStatement = new iam.PolicyStatement({
  effect: iam.Effect.ALLOW,
  actions: [''],
  resources: ['']
});

// Do some if else manipulation

statements = cdk.Fn.conditionIf(attachLambdaToVpcCondition.logicalId,
{"actions if condition is true"},
{"actions if condition is false"})

const cfnDist = policyStatement.node.defaultChild as cdk.CfnResource;
 cfnDist.addPropertyOverride("PolicyDocument", statements);

// create a lambda with a role
const myFunction = new lambda.Function(this, 'MyFunction', {
  role: myRole
});
```
But it requires more or less a major refactor that we can do in the next PR. WDYT?